### DISCUSSION TITLE
Roadmapping: handling stdout/stderr/stdin in shim

### DISCUSSION THREAD

**Discussion by robgjansen** [shadow/shadow#2728] [Show and tell]
### Current state

-  if managed process writes to STDOUT_FILENO, STDERR_FILENO with interposition disabled, it writes to shadow's stdout/stderr - i.e. terminal or shadow log.

### proposal v1
- after vfork, open /dev/null for stdout/stdin/stderr
  - any stdout/stderr output printed when interposition is disabled and before shim is initialized will be dropped instead of printed to the terminal
- after shim initializes and opens log file, set stdout/stderr to log file
  - any stdout/stderr output printed when interposition is disabled will be written in the shim log file
  - the log file isn't the best place for this, but better than printing to the terminal or dropping

### proposal v2:

- after vfork, open shim log file
- put shimlog fd in env variable, and the shim will use that fd for the shim log

### proposal v3
- vfork, open shim log file, dup to STDOUT_FILENO and STDERR_FILENO, exec
- no longer need to pass shim log name or fileno; just always use STDOUT_FILENO

**Comment by stevenengler** [comment]
This has been implemented in #2725.

**Reply by robgjansen** [reply]
I think you mean proposal v3 was implemented, right?

**Reply by stevenengler** [reply]
Yeah, but we haven't set stdin to `/dev/null` since it's less clear that it would be useful, and it would use an extra open file that would count towards the system limit.
